packages/react-fs/src/ReactFilesystem.js

```diff
@@ -0,0 +1,122 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import type {Wakeable, Thenable} from 'shared/ReactTypes';
+
+import {unstable_getCacheForType} from 'react';
+import * as fs from 'fs/promises';
+
+const Pending = 0;
+const Resolved = 1;
+const Rejected = 2;
+
+type PendingResult = {|
+  status: 0,
+  value: Wakeable,
+|};
+
+type ResolvedResult<T> = {|
+  status: 1,
+  value: T,
+|};
+
+type RejectedResult = {|
+  status: 2,
+  value: mixed,
+|};
+
+type Result<T> = PendingResult | ResolvedResult<T> | RejectedResult;
+
+function toResult<T>(thenable: Thenable<T>): Result<T> {
+  const result: Result<T> = {
+    status: Pending,
+    value: thenable,
+  };
+  thenable.then(
+    value => {
+      if (result.status === Pending) {
+        const resolvedResult = ((result: any): ResolvedResult<T>);
+        resolvedResult.status = Resolved;
+        resolvedResult.value = value;
+      }
+    },
+    err => {
+      if (result.status === Pending) {
+        const rejectedResult = ((result: any): RejectedResult);
+        rejectedResult.status = Rejected;
+        rejectedResult.value = err;
+      }
+    },
+  );
+  return result;
+}
+
+function readResult<T>(result: Result<T>): T {
+  if (result.status === Resolved) {
+    return result.value;
+  } else {
+    throw result.value;
+  }
+}
+
+function createReadFileCache(): Map<string, Result<Buffer>> {
+  return new Map();
+}
+
+export function readFile(
+  path: string,
+  options:
+    | string
+    | {
+        encoding?: string | null,
+        // Unsupported:
+        flag?: string, // Doesn't make sense except "r"
+        signal?: mixed, // We'll have our own signal
+      },
+): string | Buffer {
+  const map = unstable_getCacheForType(createReadFileCache);
+  let entry = map.get(path);
```

Comment:
Should this path be normalized to absolute paths?

I guess the same question applies to `fetch` urls?

Reply:
Is it a perf hit to normalize every time? I don't know.

Reply:
Maybe we can DEV warn if it's not already normalized?

Reply:
I think it makes sense to [`path.normalize`](https://github.com/nodejs/node/blob/8dbeac6c0480c4c85bf4c2be0bf4ddf0849f8ff2/lib/path.js#L280) to get rid of `..` and such which are pretty common if you use `__dirname` and so on. `path.normalize` is just string manipulation and should be very fast.

I don't know if it makes sense to do absolute paths via [`path.resolve`](https://github.com/nodejs/node/blob/8dbeac6c0480c4c85bf4c2be0bf4ddf0849f8ff2/lib/path.js#L138). This one may do `process.cwd()` and look at `process.env`. Not sure how those are implemented. It's unfortunate to do them all the time if the underlying functions have to do the same anyway?

Reply:
I guess even `path.resolve` has a fast path for absolute paths.

Reply:
I feel like you probably should always be resolving some absolute path in user space anyway. E.g. you can `path.resolve` in module scope and then add to that. Because relying on cwd() is fragile in general but also it can change.

It doesn't make sense to rely on what cwd() is at the time of render because that then relies on mutable state and can change between rerenders. So it doesn't make sense to rely on cwd() in our helper doing resolution but it also doesn't make sense to rely on the underlying functions reading cwd().

So I feel like we should just warn in DEV if it's not absolute.

Then the question becomes if it should also be required to already be normalized.

It seems pretty common to want to join `__dirname` with some relative path in two different files with different relative paths that ultimately resolve into the same file. In that case you'd end up reading it twice instead of deduping.

Reply:
wow I forgot cwd can change